### PR TITLE
Use ClusterMetadata based search attributes

### PR DESCRIPTION
**Pull request by alexshtin**
**What changed?**
Use ClusterMetadata based search attributes.

**Why?**
Migration of search attributes from dynamic config to main database.

**How did you test it?**
Existing and new test. Local test end to end.

**Potential risks**
Search attributes and visibility queries are not working.

**Is hotfix candidate?**
No.

### CODE REVIEW

File: proto/internal/temporal/server/api/adminservice/v1/service.proto [hunk 1]
```diff
@@ -65,8 +65,12 @@ service AdminService {
     rpc ReapplyEvents (ReapplyEventsRequest) returns (ReapplyEventsResponse) {
     }
 
-    // AddSearchAttribute whitelist search attribute in request.
-    rpc AddSearchAttribute (AddSearchAttributeRequest) returns (AddSearchAttributeResponse) {
+    // AddSearchAttribute add custom search attributes and returns comprehensive information about them.
+    rpc AddSearchAttributes (AddSearchAttributesRequest) returns (AddSearchAttributesResponse) {
```
Comment (wxing1292) [inline proto/internal/temporal/server/api/adminservice/v1/service.proto:69]: AddSearchAttribute -> AddSearchAttributes

compatibility issue?

File: temporal/server.go [hunk 2]
```diff
@@ -411,6 +416,75 @@ func verifyPersistenceCompatibleVersion(config config.Persistence, persistenceSe
 	return nil
 }
 
+// TODO: remove this func after 1.10 release
+func copyCustomSearchAttributesFromDynamicConfigToClusterMetadata(cfg *config.Config, persistenceServiceResolver resolver.ServiceResolver, logger log.Logger, dc *dynamicconfig.Collection) {
```
Comment (wxing1292) [inline temporal/server.go:420]: format?

File: temporal/server.go [hunk 3]
```diff
@@ -411,6 +416,75 @@ func verifyPersistenceCompatibleVersion(config config.Persistence, persistenceSe
...
+		logger.Error("Unable to read search attributes from dynamic config. Search attributes migration is cancelled.", tag.Error(err))
+		return
+	}
+	dcCustomSearchAttributes := searchattribute.FilterCustomOnly(dcSearchAttributes)
+	if len(dcCustomSearchAttributes) == 0 {
+		logger.Debug("Search attributes are not defined in dynamic config. Search attributes migration is cancelled.", tag.Error(err))
+		return
+	}
+
+	factory := persistenceClient.NewFactory(
```
Comment (wxing1292) [inline temporal/server.go:449]: also close this factory?

File: common/searchattribute/system_provider.go [hunk 4]
```diff
@@ -32,52 +32,25 @@ type (
...
 }
 
-// FilterCustom returns new search attributes map with only custom search attributes in it.
-func FilterCustom(searchAttributes map[string]enumspb.IndexedValueType) map[string]enumspb.IndexedValueType {
+// FilterCustomOnly returns new search attributes map with only custom search attributes in it.
+// TODO: Remove this after 1.10.0 release
+func FilterCustomOnly(searchAttributes map[string]enumspb.IndexedValueType) map[string]enumspb.IndexedValueType {
 	customSearchAttributes := map[string]enumspb.IndexedValueType{}
 	for saName, saType := range searchAttributes {
+		if saName == NamespaceID {
```
Comment (wxing1292) [inline common/searchattribute/system_provider.go:48]: why special treating the namespace ID? or should other system search attr be excluded here?

File: common/searchattribute/search_attirbute.go [hunk 5]
```diff
@@ -78,35 +78,15 @@ func BuildTypeMap(validSearchAttributesFn dynamicconfig.MapPropertyFn) (map[stri
...
 	for saName, saPayload := range searchAttributes.GetIndexedFields() {
 		_, metadataHasValueType := saPayload.Metadata[MetadataType]
 		if metadataHasValueType {
 			continue
 		}
-		valueType, isDefined := typeMap[saName]
-		if !isDefined {
-			continue
-		}
+		valueType, _ := typeMap.GetType(saName)
```
Comment (wxing1292) [inline common/searchattribute/search_attirbute.go:89]: `valueType, _ := typeMap.GetType(saName)`

what if the saName is not present?

File: common/searchattribute/name_type_map.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+	indexNameTypeMap := make(map[string]NameTypeMap, len(indexSearchAttributes))
+	for indexName, customSearchAttributes := range indexSearchAttributes {
+		indexNameTypeMap[indexName] = NameTypeMap{
+			customSearchAttributes: customSearchAttributes.GetCustomSearchAttributes(),
+		}
+	}
+	return indexNameTypeMap
+}
+
+func (m NameTypeMap) System() map[string]enumspb.IndexedValueType {
```
Comment (wxing1292) [inline common/searchattribute/name_type_map.go:50]: `func (m NameTypeMap)` -> `func (m *NameTypeMap)`

File: service/frontend/adminHandler.go [hunk 7]
```diff
@@ -156,47 +160,127 @@ func (adh *AdminHandler) AddSearchAttribute(ctx context.Context, request *admins
...
+
+	resp, err := adh.getSearchAttributes(ctx, indexName, "")
+	if err != nil {
+		return nil, adh.error(err, scope)
+	}
+
+	return resp, nil
+}
+
+func (adh *AdminHandler) getSearchAttributes(ctx context.Context, indexName, runID string) (*adminservice.GetSearchAttributesResponse, error) {
```
Comment (wxing1292) [inline service/frontend/adminHandler.go:255]: `ctx context.Context, indexName, runID string` -> `ctx context.Context, indexName string, runID string`


let us try not to omit the type of a input var
` indexName, runID string` is hard to parse, compared to ` indexName string, runID string`

File: service/frontend/adminHandler.go [hunk 8]
```diff
@@ -156,47 +160,127 @@ func (adh *AdminHandler) AddSearchAttribute(ctx context.Context, request *admins
...
+		return nil, adh.error(serviceerror.NewInternal(fmt.Sprintf(errUnableToGetSearchAttributesMessage, err)), scope)
+	}
+
+	for saName, saType := range request.GetSearchAttributes() {
+		if searchattribute.IsReservedField(saName) {
+			return nil, adh.error(serviceerror.NewInvalidArgument(fmt.Sprintf(errSearchAttributeIsReservedMessage, saName)), scope)
 		}
-		if _, exist := currentValidAttr[k]; exist {
-			return nil, adh.error(serviceerror.NewInvalidArgument(fmt.Sprintf(errKeyIsAlreadyWhitelistedMessage, k)), scope)
+		if currentSearchAttributes.IsValid(saName) {
```
Comment (wxing1292) [inline service/frontend/adminHandler.go:184]: `!if currentSearchAttributes.IsValid(saName) {` ?

File: service/frontend/adminHandler.go [hunk 9]
```diff
@@ -156,47 +160,127 @@ func (adh *AdminHandler) AddSearchAttribute(ctx context.Context, request *admins
...
-		currentValidAttr[k] = int(v)
+	// Execute workflow.
+	wfParams := addsearchattributes.WorkflowParams{
+		CustomAttributesToAdd: request.GetSearchAttributes(),
+		IndexName:             indexName,
 	}
 
-	// update dynamic config
-	err := adh.DynamicConfigClient.UpdateValue(dynamicconfig.ValidSearchAttributes, currentValidAttr)
+	run, err := adh.GetSDKClient().ExecuteWorkflow(
```
Comment (wxing1292) [inline service/frontend/adminHandler.go:198]: how long will this workflow take? 
it is not a good idea to turn a API call into a workflow execution and wait for the result

maybe just start the workflow and return the pointer?

File: service/frontend/adminHandler.go [hunk 10]
```diff
@@ -156,47 +160,127 @@ func (adh *AdminHandler) AddSearchAttribute(ctx context.Context, request *admins
...
+}
+
+func (adh *AdminHandler) getSearchAttributes(ctx context.Context, indexName, runID string) (*adminservice.GetSearchAttributesResponse, error) {
+	var lastErr error
+	descResp, err := adh.GetSDKClient().DescribeWorkflowExecution(ctx, addsearchattributes.WorkflowName, runID)
+	var wfInfo *workflowpb.WorkflowExecutionInfo
+	if err != nil {
+		// NotFound can happen when no search attributes were added and the workflow has never been executed.
+		if _, notFound := err.(*serviceerror.NotFound); !notFound {
+			lastErr = serviceerror.NewInternal(fmt.Sprintf("unable to get %s workflow state: %v", addsearchattributes.WorkflowName, err))
```
Comment (wxing1292) [inline service/frontend/adminHandler.go:262]: why not return earlier?

lastErr will also be overrided later

File: proto/internal/temporal/server/api/adminservice/v1/service.proto (hunk 1 above)
Comment (alexshtin) [reply proto/internal/temporal/server/api/adminservice/v1/service.proto:69]: Yes, It is a breaking change. I need to add this to PR description. But given that:
1. This API was always broken,
2. I had to change request and response objects too, 
I believe renaming it, is a good idea.

File: temporal/server.go (hunk 3 above)
Comment (alexshtin) [reply temporal/server.go:449]: Closing any manager created by the factory cause factory to close too. This sounds very wrong to me, I will fix it one day.

File: common/searchattribute/system_provider.go (hunk 4 above)
Comment (alexshtin) [reply common/searchattribute/system_provider.go:48]: Yes, `NamespaceId` is different. It should be hidden from user and I will add check for query parser later.

File: common/searchattribute/search_attirbute.go (hunk 5 above)
Comment (alexshtin) [reply common/searchattribute/search_attirbute.go:89]: in case of error `typeMap.GetType` returns type as `enumspb.INDEXED_VALUE_TYPE_UNSPECIFIED` which is properly handled by next call to `setMetadataType`.  I can add something like:
```
if err != nil {
    continue
}
```
because if `saName` is not defined (which should never happened under normal circumstances), there is no way to set it type.

File: common/searchattribute/name_type_map.go (hunk 6 above)
Comment (alexshtin) [reply common/searchattribute/name_type_map.go:50]: I wanted to pass `NameTypeMap` by value everywhere because it is just one pointer to the `map`. But few invocation requires `nil` because they might work w/o type map.

File: service/frontend/adminHandler.go (hunk 8 above)
Comment (alexshtin) [reply service/frontend/adminHandler.go:185]: `IsValid` is just a bad name. I am renaming it to `IsDefined` and this `if` checks if search attribute with this name is already defined.

File: service/frontend/adminHandler.go (hunk 9 above)
Comment (alexshtin) [reply service/frontend/adminHandler.go:199]: With my local tests it takes less than 1 seconds. But I agree on big ES clusters it might be longer. I would leave it as is for now because it gives better user experience. And test it with larger cluster to check if it affects run time.

File: service/frontend/adminHandler.go (hunk 10 above)
Comment (alexshtin) [reply service/frontend/adminHandler.go:263]: So this `getSearchAttributes` is called from two places `GetSearchAttributes` and `AddSearchAttributes`. For the second one, I decided not to fail entire call if get has failed (primary goal for this API is to add search attribute, not to list it). But I still want to collect as much info as I can (call to ES mapping and SDK client are independent). From other side `GetSearchAtributes` should fail if some data is not available. I guess it doesn't matter if it fails with first error or with last one.

File: common/searchattribute/search_attirbute.go (hunk 5 above)
Comment (wxing1292) [reply common/searchattribute/search_attirbute.go:89]: let us try to decrease the number of implicit contracts, 
either add a comment explaining the behavior, or make the error case / edge case explicit

File: common/searchattribute/name_type_map.go (hunk 6 above)
Comment (wxing1292) [reply common/searchattribute/name_type_map.go:50]: NameTypeMap is immutable?

File: service/frontend/adminHandler.go (hunk 10 above)
Comment (wxing1292) [reply service/frontend/adminHandler.go:263]: i would prefer not shadowing the error

if an error is not important, then log it then ignore.
if an error is important, then return it.

**Review by wxing1292 — APPROVED** [review]

File: common/searchattribute/search_attirbute.go (hunk 5 above)
Comment (alexshtin) [reply common/searchattribute/search_attirbute.go:89]: Added the snippet above.

File: common/searchattribute/name_type_map.go (hunk 6 above)
Comment (alexshtin) [reply common/searchattribute/name_type_map.go:50]: Yes, this `map` pointer is never get changed. Even when content of the map is changed, it goes through database read/write, so it is new struct every time.

File: service/frontend/adminHandler.go (hunk 10 above)
Comment (alexshtin) [reply service/frontend/adminHandler.go:263]: I reordered calls in the right priority order and added logging for every error.
